gh-105775: Convert LOAD_CLOSURE to a pseudo-op

polynomialherder opened the pull request:
Per #105775 this change converts `LOAD_CLOSURE` from an opcode to a pseudo-opcode. This enables removal of checks for uninitialized variables, and frees up an instruction.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-105775 -->
* Issue: gh-105775
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
:books: Documentation preview :books:: https://cpython-previews--106059.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

polynomialherder committed:
gh-105775: Demote LOAD_CLOSURE to a pseudo-op

This change demotes LOAD_CLOSURE from an instruction to a
pseudo-instruction. This enables superinstruction formation,
removal of checks for uninitialized variables, and frees
up an instruction.

polynomialherder committed:
gh-106044: Fix whitespace

polynomialherder committed:
gh-105775: Regenerate generated_cases.c.h

polynomialherder committed:
Merge branch 'main' into Remove-LOAD-CLOSURE-105775

carljm commented on Doc/library/dis.rst:
```diff
@@ -1725,6 +1713,18 @@ but are replaced by real opcodes or removed before bytecode is generated.
+   storage.
+
+   Note that ``LOAD_CLOSURE`` is replaced with ``LOAD_FAST`` in the assembler.
+   It exists to keep bytecode a little more readable.
```
I don't think this sentence is accurate anymore (if it ever was). `LOAD_CLOSURE` still exists in order to keep `fix_cell_offsets` working. Pseudo-instructions can't make bytecode more readable, since they don't exist in bytecode.

I would just remove this sentence.

carljm commented on Misc/NEWS.d/next/Core and Builtins/2023-06-24-10-34-27.gh-issue-105775.OqjoGV.rst:
```diff
@@ -0,0 +1,4 @@
+Remove the LOAD_CLOSURE opcode and make it a pseudo-op instead. This enables
+
+* Superinstruction formation
+* Removal of checks for uninitialized variables
```
I don't think this PR does enable either of these things. This is because `_PyCfg_OptimizeCodeUnit` is called before `_PyCfg_ConvertPseudoOps`, in `optimize_and_assemble_code_unit`. When we optimize, pseudo instructions are still present.

I think this PR is still a useful incremental step (there's no reason for `LOAD_CLOSURE` to exist as a distinct opcode in actual bytecode or in the eval loop), but actually enabling it to participate in these optimizations will require either a) refactoring how cell/local offsets are handled in the compiler so we don't need to track the distinction between `LOAD_FAST` and `LOAD_CLOSURE` through to `fix_cell_offsets`, and eliminating the `LOAD_CLOSURE` pseudo-op entirely, or b) adding explicit `LOAD_CLOSURE` support to some optimizations.

We could also experiment with trying to remove pseudo-ops before optimizing, but we'd need to audit for any dependence in the optimizer on pseudo-ops, and I think currently there'd also be a problem with `prepare_localsplus` (which calls `fix_cell_offsets`) being called after optimization, and needing `LOAD_CLOSURE`.

Correction: I guess this PR does enable "Removal of checks for uninitialized variables," simply by virtue of unconditionally converting `LOAD_CLOSURE` to `LOAD_FAST` (not `LOAD_FAST_CHECK`.)

polynomialherder replied:
Thanks, that makes sense @carljm -- I realize now that I naively copied what was in the original issue, but should have taken into account the different approach + what was happening in `_PyCompile_Assemble`

carljm commented on Python/bytecodes.c:
```diff
@@ -177,12 +176,9 @@ dummy_func(
-        inst(LOAD_CLOSURE, (-- value)) {
-            /* We keep LOAD_CLOSURE so that the bytecode stays more readable. */
-            value = GETLOCAL(oparg);
-            ERROR_IF(value == NULL, unbound_local_error);
```
I note that we are implicitly eliminating this NULL check by converting `LOAD_CLOSURE` to `LOAD_FAST`. I _think_ that this is actually just fine, because ensuring this local exists (and is a cell variable) is the responsibility of the compiler, not the code author. The compiler will always insert a `MAKE_CELL` at the top of the function for any cell variable, so I can't see any way this null check would ever have failed. If it ever did, that would indicate a compiler bug, so it's not clear that `UnboundLocalError` would ever have made sense here anyway.

gvanrossum replied:
@markshannon What do you think? We could replace it with LOAD_CLOSURE_CHECK instead to preserve this. There might be a way for a debugger to delete the variable.

carljm replied:
> There might be a way for a debugger to delete the variable.

I don't believe there is, unless the debugger is doing something unsupported and poking directly at the fastlocals array on a `_PyInterpreterFrame`. The supported way for debuggers to change the value of variables is via `frame.f_locals` dict, which is synced back to fast locals by `PyFrame_LocalsToFast`, and will only update cell contents, never overwrite or delete the cell itself.

polynomialherder committed:
gh-105775: Add unit test, correct NEWS/docs

This change adds a unit test for assembly
of code objects containing the LOAD_CLOSURE
pseud-instruction

We also correct the documentation and NEWS
descriptions to more accurately reflect the
effects of converting LOAD_CLOSURE to a pseudo-op
and the reason for its existence in the first place

polynomialherder committed:
Merge branch 'main' into Remove-LOAD-CLOSURE-105775

polynomialherder committed:
Merge branch 'Remove-LOAD-CLOSURE-105775' of github.com:polynomialherder/cpython into Remove-LOAD-CLOSURE-105775

carljm reviewed (approved):
LGTM. Thanks @polynomialherder !

polynomialherder committed:
Merge branch 'main' into Remove-LOAD-CLOSURE-105775

polynomialherder committed:
Merge branch 'main' into Remove-LOAD-CLOSURE-105775

gvanrossum commented on Python/bytecodes.c:
```diff
@@ -132,7 +132,6 @@ dummy_func(
 // BEGIN BYTECODES //
         inst(NOP, (--)) {
         }
-
```
Please put this blank line back.

gvanrossum commented on Python/bytecodes.c:
```diff
@@ -199,7 +195,6 @@ dummy_func(
             // do not use SETLOCAL here, it decrefs the old value
             GETLOCAL(oparg) = NULL;
         }
-
```
Another blank line accidentally deleted.

gvanrossum commented on Python/compile.c:
```diff
@@ -843,9 +843,10 @@ stack_effect(int opcode, int oparg, int jump)
              * of __(a)enter__ and push 2 values before jumping to the handler
              * if an exception be raised. */
             return jump ? 1 : 0;
-
```
Please keep this blank line too. I like the grouping it implies.

gvanrossum reviewed (commented):
I recommend @iritkatriel review this as well.

iritkatriel commented on Doc/library/dis.rst:
```diff
@@ -1725,6 +1713,17 @@ but are replaced by real opcodes or removed before bytecode is generated.
+   Note that ``LOAD_CLOSURE`` is replaced with ``LOAD_FAST`` in the assembler.
+
+   .. versionchanged:: 3.13
+      This opcode was demoted from full opcode to pseudo-instruction
```
```suggestion
      This opcode is now a pseudo-instruction.
```

iritkatriel commented on Lib/opcode.py:
```diff
@@ -291,6 +289,7 @@ def pseudo_op(name, op, real_ops):
 pseudo_op('LOAD_SUPER_METHOD', 263, ['LOAD_SUPER_ATTR'])
 pseudo_op('LOAD_ZERO_SUPER_METHOD', 264, ['LOAD_SUPER_ATTR'])
 pseudo_op('LOAD_ZERO_SUPER_ATTR', 265, ['LOAD_SUPER_ATTR'])
+pseudo_op('LOAD_CLOSURE', 267, ['LOAD_FAST'])
```
Let's keep them in increasing numerical order (I'm pretty sure all the other opcodes are).

iritkatriel commented on Lib/test/test_compiler_assemble.py:
```diff
@@ -70,3 +70,40 @@ def test_simple_expr(self):
+                return x
+            return inner() % 2
+
+        inner_code = mod_two.__code__.co_consts[1]
```
Add an assertion that inner_code is a code object instance (it will be easier to debug if something changes and we need a different index into co_consts).

iritkatriel commented on Misc/NEWS.d/next/Core and Builtins/2023-06-24-10-34-27.gh-issue-105775.OqjoGV.rst:
```diff
@@ -0,0 +1 @@
+Remove the LOAD_CLOSURE opcode and make it a pseudo-op instead.
```
```suggestion
Make :opcode:`LOAD_CLOSURE` a pseudo-op.
```

iritkatriel commented on Lib/importlib/_bootstrap_external.py:
```diff
@@ -451,6 +451,7 @@ def _write_atomic(path, data, mode=0o666):
 #     Python 3.13a1 3553 (Add SET_FUNCTION_ATTRIBUTE)
 #     Python 3.13a1 3554 (more efficient bytecodes for f-strings)
 #     Python 3.13a1 3555 (generate specialized opcodes metadata from bytecodes.c)
+#     Python 3.13a1 3556 (Remove LOAD_CLOSURE)
```
```suggestion
#     Python 3.13a1 3556 (Make LOAD_CLOSURE a pseudo-op)
```

iritkatriel reviewed (commented):
Let's not insult pseudo ops by calling this a demotion. Pseudo ops are awesome.

polynomialherder committed:
gh-105775: Update documentation, whitespace tweaks

- Describe the LOAD_CLOSURE change using more neutral language
- Add back in deleted whitespace
- Add an assertion to
  IsolatedAssembleTests.test_expression_with_pseudo_instruction_load_closure
- Regenerate generated files

polynomialherder committed:
Merge branch 'main' into Remove-LOAD-CLOSURE-105775

iritkatriel commented on Misc/NEWS.d/next/Core and Builtins/2023-06-24-10-34-27.gh-issue-105775.OqjoGV.rst:
```diff
@@ -0,0 +1 @@
+Convert :opcode:`LOAD_CLOSURE` to a pseudo-op.
```
```suggestion
:opcode:`LOAD_CLOSURE` is now a pseudo-op.
```

polynomialherder committed:
Update news item

polynomialherder committed:
Merge branch 'main' into Remove-LOAD-CLOSURE-105775

polynomialherder committed:
Merge branch 'main' into Remove-LOAD-CLOSURE-105775

gvanrossum reviewed (approved):
Looks good to me. Thanks for your contribution -- this is good work, and we're looking forward to your next project!
